certbot-apache/certbot_apache/_internal/configurator.py

```diff
@@ -2378,6 +2403,226 @@ def enable_autohsts(self, _unused_lineage, domains):
+            cache_path = filepath + ".db"
+            try:
+                database = bsddb.hashopen(cache_path, 'w')
+            except Exception:
+                raise errors.PluginError("Unable to open dbm database file.")
+        except ImportError:
+            # Python3 doesn't have bsddb module, so we use dbm.ndbm instead
+            import dbm
+            try:
+                database = dbm.ndbm.open(filepath, 'w')  # pylint: disable=no-member
+            except Exception:
+                # This is raised if a file cannot be found
+                raise errors.PluginError("Unable to open dbm database file.")
+        return database
+
+    def _ocsp_dbm_close(self, database):
+        """Helper method to sync and close a DBM file, in a way required by the
+        used dbm implementation."""
+        if self._ocsp_dbm_bsddb:
+            database.sync()
+            database.close()
+        else:
+            database.close()
+
+    def _ocsp_refresh_if_needed(self, pf_obj):
+        """Refreshes OCSP response for a certiifcate if it's due
+
+        :param dict pf_obj: OCSP prefetch object from pluginstorage
+
+        :returns: If OCSP response was updated
+        :rtype: bool
+
+        """
+        ttl = pf_obj["lastupdate"] + constants.OCSP_INTERNAL_TTL
+        if ttl < time.time():
+            self._ocsp_refresh(pf_obj["cert_path"], pf_obj["chain_path"])
+            return True
+        return False
+
+
+    def _ocsp_refresh(self, cert_path, chain_path):
+        """Refresh the OCSP response for a certificate
+
+        :param str cert_path: Filesystem path to certificate file
+        :param str chain_path: Filesystem path to certificate chain file
+
+        """
+
+        self._ensure_ocsp_dirs()
+        ocsp_workfile = os.path.join(
+            self.config.work_dir, "ocsp",
+            apache_util.certid_sha1_hex(cert_path))
+        handler = ocsp.RevocationChecker()
+        if not handler.revoked(cert_path, chain_path, ocsp_workfile):
+            # Guaranteed good response
+            cache_path = os.path.join(self.config.config_dir, "ocsp", "ocsp_cache")
+            # dbm.open automatically adds the file extension, it will be
+            db = self._ocsp_dbm_open(cache_path)
+            cert_sha = apache_util.certid_sha1(cert_path)
+            db[cert_sha] = self._ocsp_response_dbm(ocsp_workfile)
+            self._ocsp_dbm_close(db)
+        else:
+            logger.warning("Encountered an issue while trying to prefetch OCSP "
+                           "response for certificate: %s", cert_path)
+
+    def _ocsp_response_dbm(self, workfile):
+        """Creates a dbm entry for OCSP response data
+
+        :param str workfile: File path for raw OCSP response
+
+        :returns: OCSP response cache data that Apache can use
+        :rtype: string
+
+        """
+
+        with open(workfile, 'rb') as fh:
+            response = fh.read()
+        ttl = constants.OCSP_APACHE_TTL
```

Comment:
Shouldn't the TTL used here and in PluginStorage be influenced by the timestamps in the OCSP response? Isn't it currently possible that the OCSP response expires before the current TTL?

Reply:
This is now handled with the recent changes to `ocsp.py`. We're now reading the `nextUpdate` value from OCSP response, and creating a TTL value of half of the delta between now and then. I think this is pretty good solution that should keep the database entry untouched by Apache, but still giving enough leeway for Apache to refresh the entry in time if anything goes wrong with Certbot attempting to do so.